Fix chat template handling for OpenAI serving

JustinTong0323 opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation
Fix #8615
<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

Refactors the chat template application logic to handle cases where the tokenizer has a preprocessor with a chat template.

This change ensures that the correct chat template is applied, whether it's directly accessible through the tokenizer or via a preprocessor.

It also adds more robust error handling for incompatible tool call formats, and handles edge cases where prompt_ids is a nested list.

The template manager also now guesses the chat template from the model path and falls back to the HuggingFace default template if none is found.
<!-- Describe the changes made in this PR. -->

## Accuracy Test

<!-- If this PR affects model-side code (e.g., kernels, model architecture), please provide accuracy test results. Ref: https://docs.sglang.ai/references/accuracy_evaluation.html -->

## Benchmark & Profiling

<!-- If this PR is expected to impact performance, please provide benchmark and profiling results. Ref: https://docs.sglang.ai/references/benchmark_and_profiling.html -->

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

JustinTong0323 committed:
misc: speed up VLM CI
Enhance test configurations by adding new arguments for mm-attention-backend and cuda-graph-max-bs across multiple vision server tests.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
Fix chat template handling for OpenAI serving

Refactors the chat template application logic to handle cases where the tokenizer has a preprocessor with a chat template.

This change ensures that the correct chat template is applied, whether it's directly accessible through the tokenizer or via a preprocessor.

It also adds more robust error handling for incompatible tool call formats, and handles edge cases where prompt_ids is a nested list.

The template manager also now guesses the chat template from the model path and falls back to the HuggingFace default template if none is found.

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -200,7 +209,10 @@ def _apply_jinja_template(
-        except Exception:
+        except Exception as e:
+            logger.warning(
+                f"Error applying chat template: {e}, no worries if different tool call format detected"
```
Can we make this log a bit more formal?

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -51,6 +51,16 @@ def __init__(
     ):
         super().__init__(tokenizer_manager)
         self.template_manager = template_manager
+        if tokenizer_manager.processor is not None and hasattr(
```
Can we add a note here?

CatherineSue commented on python/sglang/srt/managers/template_manager.py:
```diff
@@ -84,26 +84,25 @@ def load_chat_template(
             self.guess_chat_template_from_model_path(model_path)
 
-            # Set default format if no template was found
+            # If pre-defined template was found, fallback to HuggingFace template
```
Ignore this, gemini already commented.

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -189,9 +199,8 @@ def _apply_jinja_template(
         try:
-            prompt_ids = self.tokenizer_manager.tokenizer.apply_chat_template(
+            prompt_ids = self.chat_template_apply_func(
                 openai_compatible_messages,
```
I'm concerned in this case, will it affect the apply_chat_template result.

JustinTong0323 commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -219,6 +231,9 @@ def _apply_jinja_template(
                 ),
             )
 
+        if isinstance(prompt_ids[0], list):
```
qwen2.5vlprocessor would return `[[id1,id2..]]` so just put it here for these kind of senario

JustinTong0323 committed:
update: revert processor.apply_caht_template

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

JustinTong0323 committed:
fix: vila

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
update

Signed-off-by: Xinyuan Tong <justinning0323@outlook.com>

CatherineSue commented on python/sglang/srt/managers/template_manager.py:
```diff
@@ -257,13 +256,20 @@ def _resolve_hf_chat_template(self, tokenizer_manager) -> Optional[str]:
+                            processor.chat_template
+                        )
+                    return processor.chat_template
+            elif tokenizer := tokenizer_manager.tokenizer:
```
```suggestion
            if tokenizer := tokenizer_manager.tokenizer:
```

JustinTong0323 committed:
fix: improve chat template handling in TemplateManager

Updated the TemplateManager to override the chat template directly from the HuggingFace template if available. Simplified the logic for retrieving the chat template from the processor and tokenizer, ensuring that the most appropriate template is applied.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
refactor: clean up unused model matching functions in conversation.py

Removed commented-out functions for various model matching templates to streamline the codebase and improve readability. This includes functions for llama2, mistral, gemma3, and others that are no longer needed.

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

JustinTong0323 committed:
update

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

CatherineSue reviewed (approved):
LGTM. If CI all passes should be good.

JustinTong0323 committed:
update audio test promp

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>
